### PR TITLE
Implemented FAN part number detection based on RPM value.

### PR DESCRIPTION
**Pull request by surf148**
When yadro-hw-manager service is started, we check availability of the
PWM properties for the fans. Then we set the PWM to 100% and after
5 secs we read the RPM value based on the obtained maximum RPM value
we detect the installed fan part number.
Note: repeatedly several times.

End-user-impact: Implemented FAN part number detection based on RPM value.
Signed-off-by: Evgeniy Kislenok <e.kislenok@yadro.com>

### CODE REVIEW

File: src/hw/hw_mngr.cpp [hunk 1]
```diff
@@ -357,3 +563,264 @@ void HWManager::setFanSpeedDelayed()
...
+        break;
+    case FanState::detect:
+        updateSystemFanFeatures();
+        detectFansState = FanState::normal;
+        break;
+    default:
+        break;
+    }
+
+    if (detectFansState == FanState::normal)
```
Comment (nest1ing) [inline src/hw/hw_mngr.cpp:745]: Why is this a separate statement?
Could we move this into `swtich-case` block?

File: src/hw/hw_mngr.cpp [hunk 2]
```diff
@@ -128,6 +262,74 @@ bool HWManager::setOption(const OptionType& optType, const int& instance,
...
+
+    //  remove existing system fans before re-publish
+    for (auto it = fans.begin(); it != fans.end(); )
+    {
+        const auto& fan = *it;
+
+        size_t pos = fan->name().find(sysFanPrefixName);
+        if (pos != std::string::npos)
+        {
+            fans.erase(it);
```
Comment (nest1ing) [inline src/hw/hw_mngr.cpp]: `it` may contain a non valid value after this call.
Likely `it = fans.erase(it);` should be here.

File: src/hw/hw_mngr.cpp [hunk 3]
```diff
@@ -234,7 +426,21 @@ void HWManager::clear()
...
+
+    //  remove all FANs except System FANs
+    for (auto it = fans.begin(); it != fans.end(); )
+    {
+        const auto& fan = *it;
+
+        size_t pos = fan->name().find(sysFanPrefixName);
+        if (pos == std::string::npos)
+        {
+            fans.erase(it);
```
Comment (nest1ing) [inline src/hw/hw_mngr.cpp]: ditto

File: src/hw/hw_mngr.cpp [hunk 4]
```diff
@@ -357,3 +563,264 @@ void HWManager::setFanSpeedDelayed()
...
+
+void HWManager::loadSystemFanFeatures()
+{
+    log<level::DEBUG>("HWManager::loadSystemFanFeatures()");
+    fanFeatures.clear();
+
+    std::ifstream dataFile(sysFanDataFile);
+    if (!dataFile.is_open())
+    {
+        log<level::DEBUG>("Cannot to open file",
```
Comment (nest1ing) [inline src/hw/hw_mngr.cpp:603]: `level::DEBUG`?
I guess we should check if file exists before an opening attempt and show this error with `level::ERROR` or even throw an exception.

File: src/hw/hw_mngr.cpp (hunk 1 above)
Comment (surf148) [reply src/hw/hw_mngr.cpp:745]: no, we can't.
if any state changes to "normal", we do publish system FAN, and this function (processDetectState) does not do any processing in normal state.

File: src/hw/hw_mngr.cpp (hunk 4 above)
Comment (surf148) [reply src/hw/hw_mngr.cpp:603]: this is not a bug. this is cache file which can be not exist. In this case, we do FAN detection.

File: src/hw/hw_mngr.cpp (hunk 3 above)
Comment (surf148) [reply src/hw/hw_mngr.cpp]: fixed

File: src/hw/hw_mngr.cpp (hunk 2 above)
Comment (surf148) [reply src/hw/hw_mngr.cpp]: fixed

File: src/hw/hw_mngr.cpp (hunk 4 above)
Comment (nest1ing) [reply src/hw/hw_mngr.cpp:603]: And what happens if the parent directory doesn't exist or is read-only?

File: src/hw/hw_mngr.cpp (hunk 3 above)
Comment (nest1ing) [reply src/hw/hw_mngr.cpp]: Where? Didn't post this yet?

File: src/hw/hw_mngr.cpp (hunk 1 above)
Comment (nest1ing) [reply src/hw/hw_mngr.cpp:745]: Oh, I've got it. This is ok.

File: src/hw/hw_mngr.cpp (hunk 4 above)
Comment (surf148) [reply src/hw/hw_mngr.cpp:603]: If we can't read file, the fanFeatures will be empty - at next time we do FAN detection and then update a file.
if we read file and format is wrong, the fanFeatures will be empty - at next time we do FAN detection and then update a file. 
if we read file and successfully parse a content, we get a not empty fanFeatures. FAN detection will be skipped.

Comment (nest1ing) [reply src/hw/hw_mngr.cpp:603]: It seems to me the file absence is only the reason that can be silently skipped, and all others should be posted to the logs with the warning/error level. But ok, let it be as is. Those other reasons are very unlikely.

**Review by nest1ing — DISMISSED** [review]

**Comment by AlexanderAmelkin** [comment]
> When yadro-hw-manager service is started, we checks available of

`we check availability of the`

> FUN's PWM-properties.

`PWM properties for the fans`

> Then we set a maximum RPM value (PWM is 100%) and after 5sec we read RPM value.

`Then we set the PWM to 100% and after 5 secs we read the RPM value`

> On RPM value we change a FAN part number.

`Based on the obtained maximum RPM value we detect the installed fan part number`

> Note: often we can't read PWM/RPM properties via dbus, therefore we
> try to read repeatly several time.

`repeatedly several times`

**Review by AlexanderAmelkin — CHANGES_REQUESTED** [review]
Please update the commit log

**Comment by AlexanderAmelkin** [comment]
:facepalm:

**Review by AlexanderAmelkin — APPROVED** [review]

**Pull request merged**